### ISSUE TITLE
Build shared library?

### ISSUE THREAD

**Issue by stuta** [litespeedtech/lsquic#137]
How to build a shared library?

cmake -DBUILD_SHARED_LIBS=ON does not create a shared library.

**Comment by dtikhonov** [comment]
There is no way to create a shared library at the moment.  There is no technical barrier to it, we just haven't had the need for it.  PRs are welcome! 🙂

**Comment by stuta** [comment]
I have some answers, but I do not know how to use cmake.

Change lsquic/src/liblsquic/CMakeLists.txt to SHARED:

```cmake
# add_library(lsquic STATIC ${lsquic_STAT_SRCS} )
add_library(lsquic SHARED ${lsquic_STAT_SRCS} )

or 

# add_library(lsquic STATIC ${lsquic_STAT_SRCS} )
IF (SHARED EQUAL 1)
	ADD_LIBRARY(lsquic SHARED ${lsquic_STAT_SRCS})
ELSE()
  ADD_LIBRARY(ls-hpack STATIC ${lsquic_STAT_SRCS})
ENDIF()
```

Build:

```bash
cmake -DBORINGSSL_DIR=/Users/xxx/installed/C/tls/boringssl -DBORINGSSL_LIB=/Users/xxx/installed/C/tls/boringssl/build -DCMAKE_BUILD_TYPE=Release -GNinja cmake ..
ninja

Errors is OSX:

```bash
Undefined symbols for architecture x86_64:
  "_ASN1_INTEGER_to_BN", referenced from:
      _lsquic_enc_session_handle_chlo in lsquic_handshake.c.o
  "_BIO_free", referenced from:
      _lsquic_bio_to_crt in lsquic_crypto.c.o
```

I tried to google those errors, but I could not find the answer.

**Comment by stuta** [comment]
I was able to generate dylib from static lib in OSX:

```bash
gcc -fpic -shared /usr/lib/libz.dylib /Users/xxx/installed/C/tls/boringssl/libcrypto.dylib /Users/xxx/installed/C/tls/boringssl/libssl.dylib /Users/xxx/installed/C/tls/boringssl/libdecrepit.dylib -Wl,-force_load, src/liblsquic/liblsquic.a -o liblsquic.dylib
```

When I tried to build shared with -DSHARED=1  in Ubuntu 20 I got this error:

```bash
[6/111] Linking C executable echo_client
FAILED: echo_client 
: && /usr/bin/cc -DLSQUIC_DEBUG_NEXT_ADV_TICK=1 -Wall -Wextra -Wno-unused-parameter -fno-omit-frame-pointer -Wno-implicit-fallthrough -Wno-missing-field-initializers -O3 -g0  -O3 -DNDEBUG  -rdynamic CMakeFiles/echo_client.dir/bin/echo_client.c.o CMakeFiles/echo_client.dir/bin/prog.c.o CMakeFiles/echo_client.dir/bin/test_common.c.o CMakeFiles/echo_client.dir/bin/test_cert.c.o  -o echo_client  -llsquic  -Wl,-Bstatic  -levent  -Wl,-Bdynamic  /home/xxx/Documents/C/boringssl/build/ssl/libssl.a  /home/xxx/Documents/C/boringssl/build/crypto/libcrypto.a  -Wl,-Bstatic  -lz  -Wl,-Bdynamic  -lpthread  -lm && :
/usr/bin/ld: cannot find -llsquic
collect2: error: ld returned 1 exit status
[8/111] Building C object CMakeFiles/echo_server.dir/bin/test_common.c.o
ninja: build stopped: subcommand failed.
```

**Comment by alandoherty** [comment]
@stuta have the same use case to generate a shared library as you, did you have any luck on Linux? I ended up with the same `sk_deep_copy` symbol problem. I managed to generate a shared library of sorts, but it was missing some things that should have been statically linked.

**Comment by alandoherty** [comment]
@stuta managed to build a really janky version of the shared library by compiling boringssl with position independent code and then adding a library target to `CMakeLists` inside `liblsquic`.Lists inside liblsquic.

```
add_library(lsquic2 SHARED ${lsquic_STAT_SRCS})
TARGET_LINK_LIBRARIES(lsquic2 PRIVATE ${BORINGSSL_LIB_ssl} ${BORINGSSL_LIB_crypto} /usr/lib/x86_64-linux-gnu/libz.so)
```

To compile boringssl so it can be statically linked into the dynamic library (not applicable to you).

```
cmake -DCMAKE_POSITION_INDEPENDENT_CODE=true .
```

**Comment by AeroNotix** [comment]
it would be very useful to have a supported way to do this as it greatly improves being able to use lsquic in FFI contexts.

**Comment by dtikhonov** [comment]
We'd take a PR...

**Comment by AeroNotix** [comment]
So I've made some changes locally that are based on what else is in this thread, can just send that up and see what you think.

**Comment by dtikhonov** [comment]
Sure, we'll check it out!

**Comment by litespeedtech** [comment]
Fixed via PR #152 -- closing.
